### PR TITLE
Remember ZfsPool state on nodes instead of double import during device scan

### PR DESCRIPTION
**Pull request by tanabarr**
**DESCRIPTION**

With ZFS MMP enabled, force import on a ZfsPool that is already import on another host fails. The implication of that in IML is that ZfsDatasets created on a ZfsPool imported on one host will not be reported by agents running on other hosts. 

This is significant because in order for a secondary mount to be identified as available for failover of a target, the agent on the secondary host of an HA pair must report that it can see the same volume (the newly created ZfsDataset created on the primary host when formatting the lustre target). With MMP enabled the force read-only import during device scan that enabled the capacity to peek the newly created ZfsDataset is no longer possible.

**SOLUTION**

After investigation and experimentation the conclusion is that because the resource manager processes storage resources dynamically, the required VolumeNodes (which identify a mapping between a host and a shared volume) are cleared as has the relevant storage resource is not reported in an agent scan.

As experience has shown, any shortcuts or hacks in the resource manager tend to have knock-on effects and generally are not with it, introducing VolumeNodes that exist without storage resources or not removing VolumeNodes in certain circumstances both result in undesired effects where objects are not cleaned up as expected.

The solution should centre around the ability to infer existence of a ZfsDataset (dataset) through its discovery on a HA pair. Meaning that the secondary host can assume a dataset exists even if the host cannot 'see' it (the host can see the ZfsPool (pool), but not the new dataset, because the pool cannot be imported) because the primary host is reporting its existence.

The resource manager could be modified to make these assumptions, that modification be a significant piece of work.

One plausible solution which may not take a large amount of effort involves adding state to the agent (not ideal) to enable it to remember datasets that it has 'seen'. This state could be stored persistently in the agent 'config store'. The result of this would be that as long as the pool had been previously imported and dataset discovered, when a pool is subsequently not importable, the datasets would continue to be reported by the agent and the relevant VolumeNodes would continue to exist. This state would be updated the next time the pool is imported.

A variation on that theme would be to share the details of datasets existing on a given pool between HA pairs by writing this information to a distributed keyvalue store which can be used to store arbitrary data. Every time a pool is imported on one of the HA pair hosts, the datasets that exist are written to the store (accessible by every host in the HA cluster). When a pool is unavailable for import (during device scan by agent's linux device plugin), the agent reports the datasets detailed in the store. 

This last variation is the recommended implementation because there are no manager changes (and therefore less undesired consequences) and only small agent changes. The distributed store will be important for future enhancements.

An interim implementation will use local storage to store last known state of zpools/datasets on an agent to enable progress whilst a suitable distributed keyvalue store is decided upon. This interim solution will work as long as pools can be exported temporarily after being formatted (datasets being created) to enable the secondary host of the HA pair to be a to 'peek' at the new dataset and therefore remember it.

---
This change is [Reviewable](https://reviewable.io/reviews/intel-hpdd/intel-manager-for-lustre/244)

### CODE REVIEW

**Comment by tanabarr** [comment]
[zfs-etcd-schema.json.txt](https://github.com/intel-hpdd/intel-manager-for-lustre/files/1268627/zfs-etcd-schema.json.txt)

suggested schema for zfs data to be held in the distributed store

**Comment by tanabarr** [comment]
still seeing pools imported between tests (which seems to be causing VolumeNodes intermittently not being created where expected and as a consequence, test failures)

**Comment by tanabarr** [comment]
now experiencing issue where node is not responding when issued with zfs/zpool commands at some point during test runs

**Comment by brianjmurrell** [comment]
> now experiencing issue where node is not responding

Is the host *oops/panic*ing perhaps?

You'd be able to determine that if we collected console output at test failure but we (a) don't collect console output at all and (b) don't collect anything (i.e. _chroma-diagnostics_/_sosreport_) at test failure.

Both of those are very worthy tickets in and of themselves.

For now, you can probably _fetch_help() on the failure and examine the state when it triggers.  Let me know if you need any help examining the state when it triggers.

File: chroma-agent/Target [hunk 1]
```diff
@@ -103,7 +103,7 @@ Target_stop() {
 }
 
 Target_status() {
-    dbg "Target_status()"
+    # dbg "Target_status()"
```
Comment (jgrund) [inline chroma-agent/Target]: Can we just remove this instead of commenting

File: chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py [hunk 2]
```diff
@@ -93,8 +96,10 @@ def test_update_conf_params(self):
         self.hosts = self.add_hosts(host_addresses)
         self.configure_power_control(host_addresses)
 
-        volumes = self.get_usable_volumes()
-        self.assertGreaterEqual(len(volumes), 4)
+        volumes = self._fetch_help(lambda: self.wait_usable_volumes(4),
```
Comment (jgrund) [inline chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py]: Does this still need to be here?

File: chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py [hunk 3]
```diff
@@ -14,7 +14,10 @@ def _create_with_params(self):
         # check for it right away and blow up if it's not as we expect.
         self.assertEqual([h['address'] for h in self.hosts], host_addresses)
 
-        volumes = self.wait_for_shared_volumes(4, 2)
+        volumes = self._fetch_help(lambda: self.wait_for_shared_volumes(4, 2),
```
Comment (jgrund) [inline chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py]: Does this still need to be here?

File: chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py [hunk 4]
```diff
@@ -151,7 +151,10 @@ def test_alerts(self):
                 self.assertEqual(alert_event['begin'], alert_event['end'])
 
             # Remove everything
-            self.graceful_teardown(self.chroma_manager)
+            self._fetch_help(lambda: self.graceful_teardown(self.chroma_manager),
```
Comment (jgrund) [inline chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py]: Does this still need to be here?

File: chroma-manager/tests/integration/core/chroma_integration_testcase.py [hunk 5]
```diff
@@ -425,12 +425,13 @@ def create_filesystem(self, hosts, filesystem, verify_successful = True):
         filesystem_id = response.json['filesystem']['id']
         command_id = response.json['command']['id']
 
-        self.wait_for_command(
-            self.chroma_manager,
-            command_id,
-            verify_successful=verify_successful,
-            timeout = LONG_TEST_TIMEOUT
-        )
+        self._fetch_help(lambda: self.wait_for_command(self.chroma_manager,
```
Comment (jgrund) [inline chroma-manager/tests/integration/core/chroma_integration_testcase.py]: Does this still need to be here?

File: chroma-manager/tests/unit/chroma_api/test_user.py [hunk 6]
```diff
@@ -76,7 +76,6 @@ def test_user_details_update(self):
         self.assertEqual(me['email'], "joebob@joebob.rocks")
         self.assertEqual(me['full_name'], "Joebob Josephson")
```
Comment (jgrund) [inline chroma-manager/tests/unit/chroma_api/test_user.py]: Can we have these changes in their own patch?

File: chroma-manager/tests/integration/core/api_testcase_with_test_reset.py [hunk 7]
```diff
@@ -943,12 +944,21 @@ def dataset_match(datasets, device_path):
...
+                                                           expected_return_code=None)
+
                         ldiskfs_device = TestBlockDevice('linux', first_test_server['device_paths'][lustre_device['path_index']])
 
                         self.execute_simultaneous_commands(ldiskfs_device.destroy_commands,
                                                            [first_test_server['fqdn']],
                                                            'clearing disk because zfs import failed %s' % ldiskfs_device,
                                                            expected_return_code=0)
+
+                        [self.remote_operations.reset_server(server['fqdn']) for server in test_servers]
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/core/api_testcase_with_test_reset.py]: Why is this necessary?  Rebooting nodes is an expensive operation in terms of test run time and should almost never be necessary.  It's a good indicator of an upstream bug if it really does need to happen.

File: chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py [hunk 8]
```diff
@@ -151,7 +151,10 @@ def test_alerts(self):
...
+                             ['tom.nabarro@intel.com'],
+                             'failure in graceful teardown with unexpected volumes',
+                             timeout=999999)
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py]: Or can we at least reduce the timeout to something smaller and more reasonable?

File: chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py (hunk 3 above)
Comment (tanabarr) [reply chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py]: will remove when confident it's not needed

File: chroma-agent/Target (hunk 1 above)
Comment (tanabarr) [reply chroma-agent/Target]: will do next repush

File: chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py (hunk 8 above)
Comment (tanabarr) [reply chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py]: have removed this one as it's no longer necessary

File: chroma-manager/tests/integration/core/api_testcase_with_test_reset.py (hunk 7 above)
Comment (tanabarr) [reply chroma-manager/tests/integration/core/api_testcase_with_test_reset.py]: this is the last resort when we cannot remove zpools normally , it is an attempt at working around #282

File: chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py (hunk 4 above)
Comment (tanabarr) [reply chroma-manager/tests/integration/shared_storage_configuration/test_alerting.py]: removed

File: chroma-manager/tests/integration/core/chroma_integration_testcase.py (hunk 5 above)
Comment (tanabarr) [reply chroma-manager/tests/integration/core/chroma_integration_testcase.py]: removed but may need to add again for further debug

File: chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py (hunk 3 above)
Comment (tanabarr) [reply chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py]: removed

File: chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py (hunk 2 above)
Comment (tanabarr) [reply chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py]: removed

File: chroma-manager/tests/integration/utils/test_blockdevices/test_blockdevice_linux.py [hunk 9]
```diff
@@ -21,6 +21,10 @@ def device_path(self):
         return self._device_path
 
     @property
+    def wipe_signatures_commands(self):
+        return ['wipefs -a %s' % self.device_path]
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/utils/test_blockdevices/test_blockdevice_linux.py]: Isn't `wipefs` a generic tool and not specific to ldiskfs (i.e. ext4)?

If so, shouldn't this be hoisted to the parent class?

File: chroma-manager/tests/integration/core/api_testcase_with_test_reset.py [hunk 10]
```diff
@@ -953,17 +953,14 @@ def dataset_match(datasets, device_path):
...
-                        #                                    expected_return_code=0)
-                        #
-                        # [self.remote_operations.reset_server(server['fqdn']) for server in test_servers]
-                        # [self.remote_operations.await_server_boot(server['fqdn']) for server in test_servers]
+                                                           'recursive destroy zpool %s' % zfs_device,
+                                                           expected_return_code=None)
+
+                        ldiskfs_device = TestBlockDevice('linux', first_test_server['device_paths'][lustre_device['path_index']])
+
+                        self.execute_simultaneous_commands(ldiskfs_device.wipe_signatures_commands,
```
Comment (brianjmurrell) [inline chroma-manager/tests/integration/core/api_testcase_with_test_reset.py]: Why are we doing a `ldiskfs` device type "wipe" on what is supposed to be a ZFS disk?

Is `cleanup_zfs_pools()` being called speculatively on devices which we don't know the contents of or is it only being called for disks which we know have ZFS on them?  I.e. because we put it there.

File: chroma-dependencies/Makefile [hunk 11]
```diff
@@ -64,6 +64,8 @@ requirements::
 	echo "systemd-python";                                    \
 	echo "iml-common>=1.0,<1.1";                              \
```
Comment (jgrund) [inline chroma-dependencies/Makefile]: Should this still be here?

File: chroma-agent/Target [hunk 12]
```diff
@@ -103,7 +103,6 @@ Target_stop() {
 }
 
 Target_status() {
-    dbg "Target_status()"
```
Comment (jgrund) [inline chroma-agent/Target]: Nice.

File: chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py [hunk 13]
```diff
@@ -17,6 +20,237 @@
...
+
+
+def write_to_store(key, value, filename=STORE_FILENAME):
+    data = {}
+
+    try:
+        with open(filename, 'r') as f:
+            data = json.loads(f.read())
+    except IOError:
+        pass
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py]: Do we want to log that we couldn't open the file?

Comment (jgrund) [reply chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py]: I'm guessing this is here due to read before write scenarios.

File: chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py [hunk 14]
```diff
@@ -17,6 +20,237 @@
...
+
+    daemon_log.info('write_to_store(): writing zfs data to %s. key: %s' % (filename, key))
+    # preserve other keys, only overwrite the key specified
+    data[key] = value
+    with open(filename, 'w') as f:
+        f.write(json.dumps(data))
+
+
+def read_from_store(key, filename=STORE_FILENAME):
+    daemon_log.info('read_from_store(): reading zfs data from %s with key: %s' % (filename, key))
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py]: Should this be `debug`?

File: chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py [hunk 15]
```diff
@@ -17,6 +20,237 @@
...
+        pass
+
+    daemon_log.info('write_to_store(): writing zfs data to %s. key: %s' % (filename, key))
+    # preserve other keys, only overwrite the key specified
+    data[key] = value
+    with open(filename, 'w') as f:
+        f.write(json.dumps(data))
+
+
+def read_from_store(key, filename=STORE_FILENAME):
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py]: Seems like this could be more general and just return the entire file.

Then you can read the key off of it in a separate call.

This could replace the same call above in `write_to_store`.

File: chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py [hunk 16]
```diff
@@ -17,6 +20,237 @@
...
+    daemon_log.info('read_from_store(): reading zfs data from %s with key: %s' % (filename, key))
+    with open(filename, 'r') as f:
+        return json.loads(f.read())[key]
+
+
+def clean_list(xs):
+    return filter_empty(strip_lines(xs))
```
Comment (jgrund) [inline chroma-agent/chroma_agent/device_plugins/linux_components/zfs.py]: This is just a compose. Could be made points-free.

```
clean_list = compose(filter_empty, strip_lines)
```

File: chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py [hunk 17]
```diff
@@ -93,8 +93,7 @@ def test_update_conf_params(self):
         self.hosts = self.add_hosts(host_addresses)
         self.configure_power_control(host_addresses)
 
-        volumes = self.get_usable_volumes()
-        self.assertGreaterEqual(len(volumes), 4)
```
Comment (jgrund) [inline chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py]: Why don't we assert here anymore?

Comment (tanabarr) [reply chroma-manager/tests/integration/shared_storage_configuration/test_conf_params.py]: assert is implicit inside wait_usable_volumes(4)

**Comment by tanabarr** [comment]
cleaned up and landed as part of #314
